flow/directives.py

```diff
@@ -0,0 +1,315 @@
+from collections.abc import MutableMapping
+import sys
+
+
+class _DirectivesItem:
+    """The representation of a single directive.
+
+    Logic for validation of values when setting, defaults, and the ability for
+    directives to interspect before returning a directive are supported. This is
+    only meant to work with the Directives class itself.
+
+    While validation is used to ensure proper setting of directives, through
+    ``finalize`` this could be abused. Also any callable can be set to any
+    directive regardless of the function taking a job as an argument and
+    returning a valid directive.
+
+    Parameters
+    ----------
+    name: str
+        The name of the directive
+    validation : (Callable[[T], S], optional)
+        A callable that accepts inputs and attempts to convert the input to a
+        valid value for the directive. If it fails, it should raise an
+        appropriate error. When not provided, the validation function just
+        returns the passed value.
+    default
+        Sets the default for the directive.
+    serial : (Callable[[S, S], S], optional)
+        A function that takes two inputs for the directive and returns the
+        appropriate value for these operations running in serial. Defaults to
+        the maximum of the two.
+    parallel : (Callable[[S, S], S], optional)
+        A function that takes two inputs for the directive and returns the
+        appropriate value for these operations running in parallel. Defaults to
+        the sum of the two.
+    finalize : (Callable[[S, Directives], S], optional):
+        A function that takes the current value of the directive and the
+        directives object it is a child of and outputs the finalized value for
+        that directive. This is useful if some directives have multiple ways to
+        be set or are dependent in some way on other directives. The default is
+        to just return the current set value.
+    """
+    def __init__(self, name, validation=None, default=None,
```

Comment:
Everything after `name` could become [keyword-only arguments](https://www.python.org/dev/peps/pep-3102/). I couldn't understand the code later in this PR where the parameters are provided without their corresponding keywords. This feature has been around for all of Python 3, though I've never used keyword-only arguments. Forcing keywords also ensures that you can re-order or add new parameters after the `*` in the API without breaking downstream callers' code.
```suggestion
    def __init__(self, name, *, validation=None, default=None,
```